keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,228 @@
+* Streamline the tedious processes involved in debugging latency issues in Kubernetes
+* Make it possible to identify high-level latency regressions, and attribute them back to offending processes
+
+
+### Non-Goals
+
+* Replace existing logging, metrics, or the events API
+* Trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+
+## Proposal
+
+### Architecture
+
+#### Tracing API Requests
+
+In the traditional tracing model, a client sends a request to a server and recieves a response back.  Even though Kubernetes "controllers" don't follow this model (more on that later), the kube-apierver and backing storage (e.g. etcd3) do.  To enable traces to be collected for API requests, the following must be true:
+
+1. The apiserver must propagate the http context of incoming requests through its function stack to the backing storage
+1. Kubernetes client libraries must allow passing a context with API requests
+
+To actually add traces to API requests, owners of the kube-apiserver and backing storage may add Spans to incoming requests, and configure sampling as they see fit.
+
+#### Propagating Context Through Objects
+
+While API requests follow the traditional RPC client-server tracing model, kubernetes controllers don't.  Instead of controller actions being driven by incoming RPCs, their actions are driven by observations of desired and actual state.  This is the primary reason why the kubernetes community hasn't agreed on how to integrate tracing into kubernetes thus far.
+
+In the traditional RPC client-server tracing model, a trace context is attached to a single incoming request, and is propagated with all requests the server makes to other servers required to fulfill the initial single request.  Conceptually, this proposal suggests treating a kubernetes cluster as a single RPC server.  The difference is that we attach context to objects, and propagate this context to objects modified as a result of the initial object modification.  For example, if a user creates a ReplicaSet, the kube-controller-manager will create many Pod objects as a result, and will propagate the context used to create the ReplicaSet to Pod objects as well.  This ensures that all actions taken by kubernetes controllers as a result of the initial user action are linked by the same context.
+
+For the alpha phase, we choose to propagate this span context as an encoded string an object annotation called `trace.kubernetes.io/context`.  As noted in [Tracing API Requests](#tracing-api-requests) above, storing the trace context with the context is _in addition_ to attaching a context to http requests to the apiserver.  The reason for this is explained in the [Controller Behavior](#controller-behavior) section below.  In some scenarios, controllers will want to update the trace context from A -> B, but want to associated that Update request with context A.
```

Comment:
Around 32 bytes (16 bytes for the trace ID, 8 bytes for the span ID, and some metadata). Here is the [OpenCensus spec](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/Span.md#spancontext) on this (not sure how/if this will be affected with OpenTelemetry), and here is the [w3c spec](https://w3c.github.io/trace-context/#tracestate-field).

Reply:
Added this to the proposal